modules/Bio/EnsEMBL/Hive/DBSQL/AnalysisJobAdaptor.pm

```diff
@@ -587,6 +587,61 @@ sub reset_or_grab_job_by_dbID {
     return $job;
 }
 
+=head2 reset_or_grab_job_by_inputID
+
+  Arg [1]    : int $analysis_id
+  Arg [2]    : string $input_id
+  Arg [3]    : int $role_id (optional)
+  Description: resets a job to to 'READY' (if no $role_id given) or directly to 'CLAIMED' so it can be run again, and fetches it.
+               NB: Will also reset a previously 'SEMAPHORED' job to READY.
+               The retry_count will be set to 1 for previously run jobs (partially or wholly) to trigger PRE_CLEANUP for them,
+               but will not change retry_count if a job has never *really* started. $input_id can be a wildcard entry.
+  Returntype : Bio::EnsEMBL::Hive::AnalysisJob or undef
+
+=cut
+
+sub reset_or_grab_job_by_inputID {
+    my ($self, $analysis_id, $input_id, $role_id) = @_;
+
+    my $new_status  = $role_id ? 'CLAIMED' : 'READY';
+
+        # Note: the order of the fields being updated is critical!
+    my $sql = qq{
+        UPDATE job
+           SET retry_count = CASE WHEN (status='READY' OR status='CLAIMED') THEN retry_count ELSE 1 END
+             , status=?
+             , role_id=?
+         WHERE input_id LIKE ? AND analysis_id=?
+    };
+    my @values = ($new_status, $role_id, $input_id, $analysis_id);
+
+    my $sth = $self->prepare( $sql );
+    my $return_code = $sth->execute( @values )
+        or die "Could not run\n\t$sql\nwith data:\n\t(".join(',', @values).')';
+    $sth->finish;
+
+
+    # Get list of input_id for given wildcard argument
+    my $sql_get_input_id = qq{
+        SELECT input_id FROM job
+         WHERE input_id LIKE ? AND analysis_id=?
```

Comment:
This won't work because input_id is subject to overflow. That is to say that the input_id value might be stored in the input_id column, or it might be stored in the analysis_data table. If the input_id is stored in analysis_data, then input_id will have a value like "_extended_data_id N" (where N is an arbitrary integer). See fetch_by_analysis_id_and_input_id for an example.

One approach here would be to fetch_all_by_analysis_id, which will get a list of relevant jobs. These jobs can then be filtered by input_id. If you call input_id on an AnalysisJob object, the value will be correct. BaseAdaptor automatically checks to see if the input_id is stored in the input_id column or has been overflow-ed to analysis_data and fetches it appropriately. Eventually, you will get a set of jobs, and you can reset them by job_id.

Reply:
Fetching all jobs is going to be a killer: there can be millions of jobs. If you want to go this way, it might be worth it to add a new `fetch_iterator` in DBSQL::BaseAdaptor and process them on the fly.
Another thing: job parameters can be available through their "stack" too (`accu_id_stack`, `param_id_stack`)

Reply:
The issue of job count is difficult to get around due to overflow. I'm not coming up with a better idea than an iterator, which at least solves the memory overhead of storing potentially millions of jobs. A view joining analysis_data will put even more load on the DB.

As for the second part, I'm inclined to have this feature just scan input_id in the first instance. It would be more useful to be able to select jobs based on any visible parameter. Adding that feature would pair nicely with adding a parameter inspector for jobs, so I'm inclined to add that in a second iteration.

Reply:
Re. the stack-inherited parameters: you'll have to be clear in the documentation that the feature doesn't work on pipelines that have the stack enabled at the pipeline level, or on analyses that follow `INPUT_PLUS`.